input-explainer.md

```diff
@@ -280,6 +294,91 @@ When the canvas receives a `pointerdown` event an `XRInputSource` is created wit
+
+A tracked controller reports the state of several input types that are commonly found on XR controllers. These are:
+
+ - Joystick
+ - Touchpad
+ - Trigger
+ - Grip button
+
+The `XRTrackedController` also exposes a list of "Extended" inputs to capture controller elements who's semantic meanings are not captured in the list above, such as face buttons.
+
+Aside from the `extended` list it is assumed that each controller either has zero or one of the above inputs. If the controller does not have one of the inputs listed above the corresponding attribute in the `controller` object must be `null`.
+
+Each input, if present, is represented as an `XRTrackedControllerState` instance. How the  values are interpreted depends on the inputs in question, but there are common guidelines to their use.
+
+  * `pressed` is a boolean that indicates if the input is physically pressed down. If the input cannot detect when it is pressed this value must be false. If the input represents an analog button or trigger `pressed` should be `true` if the analog value is greater than a system-recommended threshold.
+  * `touched` is a boolean that indicates if the input is currently being touched by the user. If the input is not capacitive (cannot detect touch) this value should mirror the `pressed` value. If the input represents an analog button or trigger `touched` should be `true` any time the analog value is non-zero.
+  * `value` is an array of doubles that represents the analog value(s) of the input. If the input represents an analog button or trigger `values` should have a single element who's value represents the full range of input normalized to a `[0, 1]` range, with `1` meaning "100% engaged". If the input represents a joystick, touchpad, or other logical group of axes `values` should have one element for each axis and report their movement range normalized to a `[-1, 1]` range, with `0` representing the neutral position of the axis. If the input is a touchpad all `value`s should report `0` when the user is not touching the input.
+
+Each input also has a `name` which is a human readable label that should correspond to the name the platform gives for the input if possible. For example, inputs in the `extended` array may have names like "A", "B", "X", and "Y". If a platform-specific name cannot be determined the name should reflect the input's semantic type, such as "Touchpad" or "Button 0".
+
+```js
+function onXRFrame(timestamp, frame) {
+  let inputSource = xrSession.getInputSources()[0];
+
+  // Check to see if the input source is a controller.
+  if (inputSource.controller) {
+    
+    // Determine if the controller has a joystick with at least 2 axes.
+    let joystick = inputSource.controller.joystick;
+    if (joystick && joystick.value.length >= 2) {
+      MoveUser(joystick.value[0], joystick.value[1]);
+    }
+
+    // Determine if the controller has at least one extended input, such as a
+    // face button.
```

Comment:
If we pursue an `extended` approach, we'll need to explicitly think through how buttons/axes will transition from `extended` to full named attributes.  Otherwise, promoting a button or axis will break existing content.

It may be more compatible to change `extended` to `elements` or such, including both the core and extended elements.  That way, there is no temptation for UAs to remove elements as they are promoted to core.

Reply:
From the call, it sounds like there are concerns about `element`, `component` or `part`.  However, we should agree on the structure we determine that we need here, and then find the least-bad name.

Reply:
I'm open to renaming `extended` to `all` or something similar with the appropriate behavior.  
That said, IIRC, webidl doesn't allow for a dictionary of values as an attribute.  At one point we considered a function like `getInputPart(string partName)`.  Perhaps that's an option?